File: rust/src/leetcode/Q79_word_search.rs
/*
 * @lc app=leetcode id=79 lang=rust
 *
 * [79] Word Search
 */

// @lc code=start
impl Solution {
    fn dfs(
        board: &Vec<Vec<char>>,
        word: &Vec<char>,
        index: usize,
        visited: &mut Vec<Vec<bool>>,
        row: i32,
        col: i32,
    ) -> bool {
        if index == word.len() {
            return true;
        }

        if row < 0
            || col < 0
            || row >= board.len() as i32
            || col >= board[0].len() as i32
            || visited[row as usize][col as usize]
            || board[row as usize][col as usize] != word[index]
        {
            return false;
        }

        visited[row as usize][col as usize] = true;

        if Self::dfs(board, word, index + 1, visited, row + 1, col) {
            return true;
        }
        if Self::dfs(board, word, index + 1, visited, row - 1, col) {
            return true;
        }
        if Self::dfs(board, word, index + 1, visited, row, col + 1) {
            return true;
        }
        if Self::dfs(board, word, index + 1, visited, row, col - 1) {
            return true;
        }

        visited[row as usize][col as usize] = false;

        false
    }

    pub fn exist(board: Vec<Vec<char>>, word: String) -> bool {
        let word: Vec<char> = word.chars().collect();
        let mut visited: Vec<Vec<bool>> = vec![vec![false; board[0].len()]; board.len()];

        for i in 0..board.len() {
            for j in 0..board[0].len() {
                if Self::dfs(&board, &word, 0, &mut visited, i as i32, j as i32) {
                    return true;
                }
            }
        }

        false
    }
}
// @lc code=end

File: python/Q79_word_search.py
#
# @lc app=leetcode id=79 lang=python3
#
# [79] Word Search
#


# @lc code=start
class Solution:
    def search(self, row: int, col: int, index: int) -> bool:
        if index == self.target:
            return True

        # out of bounds, visited cell, wrong letter
        if (
            row < 0
            or col < 0
            or row >= self.ROWS
            or col >= self.COLS
            or self.visiting[row][col]
            or self.board[row][col] != self.word[index]
        ):
            return False

        self.visiting[row][col] = True
        result: bool = any(
            self.search(row + d1, col + d2, index + 1) for d1, d2 in self.directions
        )
        self.visiting[row][col] = False

        return result

    def exist(self, board: list[list[str]], word: str) -> bool:
        self.ROWS: int = len(board)
        self.COLS: int = len(board[0])

        self.board: list[list[str]] = board
        self.word: str = word
        self.target: int = len(self.word)

        self.directions: list[tuple[int, int]] = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        self.visiting: list[list[bool]] = [
            [False for _ in range(self.COLS)] for _ in range(self.ROWS)
        ]

        for i in range(self.ROWS):
            for j in range(self.COLS):
                if self.search(i, j, 0):
                    return True

        return False


# @lc code=end
